nixos/modules/services/desktop-managers/cosmic.nix

```diff
@@ -0,0 +1,166 @@
+    services.accounts-daemon.enable = true;
+    services.displayManager.sessionPackages = with pkgs; [ cosmic-session ];
+    services.libinput.enable = true;
+    services.upower.enable = true;
+    systemd.packages = with pkgs; [ cosmic-session ];
+    xdg.icons.enable = true;
+    xdg.mime.enable = true;
+
+    # Good to have options
+    hardware.bluetooth.enable = lib.mkDefault true;
+    networking.networkmanager.enable = lib.mkDefault true;
+    programs.dconf.enable = lib.mkDefault true;
+    services.acpid.enable = lib.mkDefault true;
+    services.gnome.gnome-keyring.enable = lib.mkDefault true;
+    services.gvfs.enable = lib.mkDefault true;
+    services.pipewire.alsa.enable = lib.mkDefault true;
+    services.pipewire.enable = lib.mkDefault true;
+    services.pipewire.pulse.enable = lib.mkDefault true;
+    services.power-profiles-daemon.enable = lib.mkDefault (!config.hardware.system76.power-daemon.enable);
+
+    warnings =
+      lib.optional
+        (
+          lib.elem pkgs.cosmic-files config.environment.cosmic.excludePackages
+          && !(lib.elem pkgs.cosmic-session config.environment.cosmic.excludePackages)
+        )
+        ''
+          The COSMIC session may fail to initialise with the `cosmic-files` package excluded via
+          `config.environment.cosmic.excludePackages`.
+
+          Please do one of the following:
+            1. Remove `cosmic-files` from `config.environment.cosmic.excludePackages`.
+            2. Add `cosmic-session` (in addition to `cosmic-files`) to
+               `config.environment.cosmic.excludePackages` and ensure whatever session starter/manager you are
+               using is appropriately set up.
+        '';
+
+    assertions = [
+      {
+        assertion = lib.elem "libcosmic-app-hook" (
+          lib.map (
+            drv: lib.optionalString (lib.isDerivation drv) (lib.getName drv)
+          ) pkgs.cosmic-comp.nativeBuildInputs
+        );
+        message = ''
+          It looks like the provided `pkgs` to the NixOS COSMIC module is not usable for a working COSMIC
+          desktop environment.
+
+          If you are erroneously passing in `pkgs` to `specialArgs` somewhere in your system configuration,
+          this is is often unnecessary and has unintended consequences for all NixOS modules. Please either
+          remove that in favor of configuring the NixOS `pkgs` instance via `nixpkgs.config` and
+          `nixpkgs.overlays`.
+
+          If you must instantiate your own `pkgs`, then please include the overlay from the NixOS COSMIC flake
+          when instantiating `pkgs` and be aware that the `nixpkgs.config` and `nixpkgs.overlays` options will
+          not function for any NixOS modules.
+
+          Note that the COSMIC packages in Nixpkgs are still largely broken as of 2024-10-16 and will not be
+          usable for having a fully functional COSMIC desktop environment. The overlay is therefore necessary.
+        '';
```

Comment:
```suggestion
```
This isn't applicable in Nixpkgs. It's also up to you to fix the bugs that this assertion tries to warn users about.